x/auth/tx/service.go

```diff
@@ -37,6 +39,83 @@ func NewTxServer(clientCtx client.Context, simulate baseAppSimulateFn, interface
 
 var _ txtypes.ServiceServer = txServer{}
 
+const (
+	eventFormat = "{eventType}.{eventAttribute}={value}"
+)
+
+// TxsByEvents implements the ServiceServer.TxsByEvents RPC method.
+func (s txServer) TxsByEvents(ctx context.Context, req *txtypes.GetTxsEventRequest) (*txtypes.TxsByEventsResponse, error) {
+
+	if req.Page < 0 {
+		return nil, status.Error(codes.InvalidArgument, "page must greater than 0")
+	}
+	if req.Limit < 0 {
+		return nil, status.Error(codes.InvalidArgument, "limit must greater than 0")
+	}
+	if len(req.Event) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "must declare at least one event to search")
+	}
+	var events []string
+	if strings.Contains(req.Event, "&") {
+		events = strings.Split(req.Event, "&")
+	} else {
+		events = append(events, req.Event)
+	}
+
+	tmEvents := make([]string, len(events))
+
+	for i, event := range events {
+		if !strings.Contains(event, "=") {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid event; event %s should be of the format: %s", event, eventFormat))
+		} else if strings.Count(event, "=") > 1 {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid event; event %s should be of the format: %s", event, eventFormat))
+		}
+
+		tokens := strings.Split(event, "=")
+		if tokens[0] == tmtypes.TxHeightKey {
+			event = fmt.Sprintf("%s=%s", tokens[0], tokens[1])
+		} else {
+			event = fmt.Sprintf("%s='%s'", tokens[0], tokens[1])
+		}
+
+		tmEvents[i] = event
+	}
+
+	query := strings.Join(tmEvents, " AND ")
+	page := int(req.Page)
+	limit := int(req.Limit)
+
+	result, err := s.clientCtx.Client.TxSearch(ctx, query, false, &page, &limit, "")
+	if err != nil {
+		return nil, err
+	}
+	// Create a proto codec, we need it to unmarshal the tx bytes.
+	cdc := codec.NewProtoCodec(s.clientCtx.InterfaceRegistry)
+	res := make([]*txtypes.TxResponse, result.TotalCount)
+	var protoTx txtypes.Tx
+	for i, tx := range result.Txs {
+		if err := cdc.UnmarshalBinaryBare(tx.Tx, &protoTx); err != nil {
+			return nil, err
+		}
+		res[i] = &txtypes.TxResponse{
```

Comment:
this logic appears twice in this file. It would be nice to factor into a function `txResponseToTxRes` (converts abci TxResponse to our TxRes)